### ISSUE TITLE
Events with short condition/action strings cause memory corruption

### ISSUE THREAD

**Issue by iamthad** [nasa/trick#587]
tl;dr On C++ standard libraries with small-string optimization, short event/condition strings can end up pointing into freed memory. We shouldn't move `std::string` with `memcpy`.

----

Trick's events use the Trick memory manager to handle arrays of condition and action objects, which contain `std::string` members. When these arrays are resized, their data is moved around using `memcpy`. Moving instances of `std::string` with `memcpy` isn't guaranteed to be safe, and it causes undefined behavior (read-after-free) when small string optimization is in place.

When the string is short enough (16 characters in libstdc++, 24 characters in libc++), it is stored within a buffer in the `std::string` instance itself, and the data pointer points at that internal buffer. If that instance is moved in memory without updating the pointer, the pointer still refers to the old memory location, which has been freed.

----

To reproduce this, apply the following patch and then run `./S_main_Linux_5.4_x86_64.exe RUN_test/events.py`. The new event action is duplicated several times to increase the chance that the freed memory is corrupted.

```diff
--- i/trick_sims/SIM_Ball++_L1/RUN_test/events.py
+++ w/trick_sims/SIM_Ball++_L1/RUN_test/events.py
@@ -11,7 +11,10 @@
 my_event.set_cycle(10.0)
 my_event.condition(0, "ball.obj.state.output.position[0] > 0.0")
-my_event.action(0 , """
+my_event.action(0, "print 'hi'")
+my_event.action(1, "print 'hi'")
+my_event.action(2, "print 'hi'")
+my_event.action(3 , """
 print "hello world" , ball.obj.state.output.position[0] , trick.exec_get_sim_time()
 my_event.activate()
 """)
```

The output will look something like 
```
  File "<string>", line 1
    print 'h
            ^
SyntaxError: EOL while scanning string literal
```

----

Alternatively, apply the following patch:
```diff
diff --git i/trick_sims/SIM_Ball++_L1/RUN_test/events.py w/trick_sims/SIM_Ball++_L1/RUN_test/events.py
index 806cdf06..00969d51 100644
--- i/trick_sims/SIM_Ball++_L1/RUN_test/events.py
+++ w/trick_sims/SIM_Ball++_L1/RUN_test/events.py
@@ -9,7 +9,10 @@
 my_event = trick.new_event("my_event")
 my_event.set_cycle(10.0)
-my_event.condition(0, "ball.obj.state.output.position[0] > 0.0")
+my_event.condition(0, "True")
+my_event.condition(1, "True")
+my_event.condition(2, "True")
+my_event.condition(3, "ball.obj.state.output.position[0] > 0.0")
 my_event.action(0 , """
 print "hello world" , ball.obj.state.output.position[0] , trick.exec_get_sim_time()
```

The errors will look something like
```
  File "<string>", line 1
    trick_ip.ip.return_val = .5e
                             ^
SyntaxError: invalid syntax
  File "<string>", line 1
    trick_ip.ip.return_val =
                            ^
SyntaxError: invalid syntax
```
This demonstrates that both conditions and events are affected.

**Comment by iamthad** [comment]
I forgot to mention something important: this only shows up when using a C++ standard library that performs small-string optimization. The library shipped with G++ version < 5 doesn't do this, and the one shipped with 5 and up won't if `_GLIBCXX_USE_CXX11_ABI` isn't enabled.

**Comment by jmpenn** [comment]
Thanks


On Mar 26, 2018, at 9:36 AM, iamthad <notifications@github.com<mailto:notifications@github.com>> wrote:


tl;dr On C++ standard libraries with small-string optimization, short event/condition strings can end up pointing into freed memory. We shouldn't move std::string with memcpy.

________________________________

Trick's events use the Trick memory manager to handle arrays of condition and action objects, which contain std::string members. When these arrays are resized, their data is moved around using memcpy. Moving instances of std::string with memcpy isn't guaranteed to be safe, and it causes undefined behavior (read-after-free) when small string optimization is in place.

When the string is short enough (16 characters in libstdc++, 24 characters in libc++), it is stored within a buffer in the std::string instance itself, and the data pointer points at that internal buffer. If that instance is moved in memory without updating the pointer, the pointer still refers to the old memory location, which has been freed.

________________________________

To reproduce this, apply the following patch and then run ./S_main_Linux_5.4_x86_64.exe RUN_test/events.py. The new event action is duplicated several times to increase the chance that the freed memory is corrupted.

--- i/trick_sims/SIM_Ball++_L1/RUN_test/events.py
+++ w/trick_sims/SIM_Ball++_L1/RUN_test/events.py
@@ -11,7 +11,10 @@
 my_event.set_cycle(10.0)
 my_event.condition(0, "ball.obj.state.output.position[0] > 0.0")
-my_event.action(0 , """
+my_event.action(0, "print 'hi'")
+my_event.action(1, "print 'hi'")
+my_event.action(2, "print 'hi'")
+my_event.action(3 , """
 print "hello world" , ball.obj.state.output.position[0] , trick.exec_get_sim_time()
 my_event.activate()
 """)

The output will look something like

  File "<string>", line 1
    print 'h
            ^
SyntaxError: EOL while scanning string literal


________________________________

Alternatively, apply the following patch:

diff --git i/trick_sims/SIM_Ball++_L1/RUN_test/events.py w/trick_sims/SIM_Ball++_L1/RUN_test/events.py
index 806cdf06..00969d51 100644
--- i/trick_sims/SIM_Ball++_L1/RUN_test/events.py
+++ w/trick_sims/SIM_Ball++_L1/RUN_test/events.py
@@ -9,7 +9,10 @@
 my_event = trick.new_event("my_event")
 my_event.set_cycle(10.0)
-my_event.condition(0, "ball.obj.state.output.position[0] > 0.0")
+my_event.condition(0, "True")
+my_event.condition(1, "True")
+my_event.condition(2, "True")
+my_event.condition(3, "ball.obj.state.output.position[0] > 0.0")
 my_event.action(0 , """
 print "hello world" , ball.obj.state.output.position[0] , trick.exec_get_sim_time()

The errors will look something like

  File "<string>", line 1
    trick_ip.ip.return_val = .5e
                             ^
SyntaxError: invalid syntax
  File "<string>", line 1
    trick_ip.ip.return_val =
                            ^
SyntaxError: invalid syntax


This demonstrates that both conditions and events are affected.
